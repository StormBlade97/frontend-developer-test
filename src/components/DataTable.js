import * as React from 'react';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableContainer from '@material-ui/core/TableContainer';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import TableSortLabel from '@material-ui/core/TableSortLabel';
import Paper from '@material-ui/core/Paper';

const DataTable = ({ columns, rows, tableController, sortBy, sortDirection, toggleSort }) => {
    const hasData = rows && rows.length > 0;
    
    return <TableContainer component={Paper}>
        <Table aria-label="simple table">
            <TableHead>
                <TableRow>
                    {
                        columns.map(header =>
                            <TableCell component="th" key={header}>
                                <TableSortLabel
                                    active={sortBy === header}
                                    direction={sortDirection}
                                    onClick={toggleSort}
                                >
                                    {header}
                                </TableSortLabel>
                             </TableCell>)
                    }
                </TableRow>
            </TableHead>
            <TableBody>
                {hasData && rows.map((row, idx) => (
                    <TableRow key={idx}>
                        <TableCell>{row.date}</TableCell>
                        <TableCell>
                            {row.userid}
                        </TableCell>
                        <TableCell>{row.oldVal}</TableCell>
                        <TableCell>{row.newVal}</TableCell>
                    </TableRow>
                ))}
                <TableRow>
                    <TableCell align="center" colSpan={columns.length}>
                        {
                            tableController
                        }
                    </TableCell>
                </TableRow>
            </TableBody>
        </Table>
    </TableContainer>;
}

export default DataTable;