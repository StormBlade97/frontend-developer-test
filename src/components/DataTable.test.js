import React from 'react';
import { shallow } from 'enzyme';
import DataTable from './DataTable';
import TableSortLabel from '@material-ui/core/TableSortLabel';

describe("DataTable", () => {
    const rows = [{
        userid: '1',
        date: '2020-12-01',
        oldVal: 'old',
        newVal: 'new'
    }, {
        userid: '2',
        date: '2020-12-02',
        oldVal: 'old',
        newVal: 'new'
        }]
    const columns = ['a', 'b', 'c', 'd']
    describe("render", () => {
        it("renders", () => {
            const wrapper = shallow(<DataTable
                tableController={"Controller"}
                columns={columns}
                sortBy={'a'}
                sortDirection={'asc'}
                rows={rows}
            />)
            expect(wrapper.debug()).toMatchSnapshot()
        })
    })
    describe("sort feature", () => {
        it("renders desc sort arrow", () => {
            const wrapper = shallow(<DataTable
                tableController={"Controller"}
                columns={columns}
                sortBy={'a'}
                sortDirection={'desc'}
                rows={rows}
            />)
            expect(wrapper.find(TableSortLabel).debug()).toMatchSnapshot();
        })
        it("renders asc sort arrow", () => {
            const wrapper = shallow(<DataTable
                tableController={"Controller"}
                columns={columns}
                sortBy={'a'}
                sortDirection={'desc'}
                rows={rows}
            />);
            expect(wrapper.find(TableSortLabel).debug()).toMatchSnapshot();
        })
        it("pass toggleSort callback to underlying TableSortLabel", () => {
            const toggleSort = jest.fn()
            shallow(<DataTable
                tableController={"Controller"}
                columns={columns}
                sortBy={'a'}
                sortDirection={'desc'}
                rows={rows}
                toggleSort={toggleSort}
            />);
            expect(toggleSort.mock.calls).toMatchSnapshot();
        })
    })
})