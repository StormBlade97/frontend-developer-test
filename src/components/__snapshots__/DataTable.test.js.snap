// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable render renders 1`] = `
"<WithStyles(ForwardRef(TableContainer)) component={{...}}>
  <WithStyles(ForwardRef(Table)) aria-label=\\"simple table\\">
    <WithStyles(ForwardRef(TableHead))>
      <WithStyles(ForwardRef(TableRow))>
        <WithStyles(ForwardRef(TableCell)) component=\\"th\\">
          <WithStyles(ForwardRef(TableSortLabel)) active={true} direction=\\"asc\\" onClick={[undefined]}>
            a
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell)) component=\\"th\\">
          <WithStyles(ForwardRef(TableSortLabel)) active={false} direction=\\"asc\\" onClick={[undefined]}>
            b
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell)) component=\\"th\\">
          <WithStyles(ForwardRef(TableSortLabel)) active={false} direction=\\"asc\\" onClick={[undefined]}>
            c
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell)) component=\\"th\\">
          <WithStyles(ForwardRef(TableSortLabel)) active={false} direction=\\"asc\\" onClick={[undefined]}>
            d
          </WithStyles(ForwardRef(TableSortLabel))>
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableHead))>
    <WithStyles(ForwardRef(TableBody))>
      <WithStyles(ForwardRef(TableRow))>
        <WithStyles(ForwardRef(TableCell))>
          2020-12-01
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          1
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          old
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          new
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableRow))>
        <WithStyles(ForwardRef(TableCell))>
          2020-12-02
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          2
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          old
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          new
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableRow))>
        <WithStyles(ForwardRef(TableCell)) align=\\"center\\" colSpan={4}>
          Controller
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
    </WithStyles(ForwardRef(TableBody))>
  </WithStyles(ForwardRef(Table))>
</WithStyles(ForwardRef(TableContainer))>"
`;

exports[`DataTable sort feature pass toggleSort callback to underlying TableSortLabel 1`] = `Array []`;

exports[`DataTable sort feature renders asc sort arrow 1`] = `
"<WithStyles(ForwardRef(TableSortLabel)) active={true} direction=\\"desc\\" onClick={[undefined]}>
  a
</WithStyles(ForwardRef(TableSortLabel))>


<WithStyles(ForwardRef(TableSortLabel)) active={false} direction=\\"desc\\" onClick={[undefined]}>
  b
</WithStyles(ForwardRef(TableSortLabel))>


<WithStyles(ForwardRef(TableSortLabel)) active={false} direction=\\"desc\\" onClick={[undefined]}>
  c
</WithStyles(ForwardRef(TableSortLabel))>


<WithStyles(ForwardRef(TableSortLabel)) active={false} direction=\\"desc\\" onClick={[undefined]}>
  d
</WithStyles(ForwardRef(TableSortLabel))>"
`;

exports[`DataTable sort feature renders desc sort arrow 1`] = `
"<WithStyles(ForwardRef(TableSortLabel)) active={true} direction=\\"desc\\" onClick={[undefined]}>
  a
</WithStyles(ForwardRef(TableSortLabel))>


<WithStyles(ForwardRef(TableSortLabel)) active={false} direction=\\"desc\\" onClick={[undefined]}>
  b
</WithStyles(ForwardRef(TableSortLabel))>


<WithStyles(ForwardRef(TableSortLabel)) active={false} direction=\\"desc\\" onClick={[undefined]}>
  c
</WithStyles(ForwardRef(TableSortLabel))>


<WithStyles(ForwardRef(TableSortLabel)) active={false} direction=\\"desc\\" onClick={[undefined]}>
  d
</WithStyles(ForwardRef(TableSortLabel))>"
`;
